handler/schedule.go

```diff
@@ -125,3 +137,65 @@ func (s *ScheduleHandler) mergeResourcesWithDefaults(resources corev1.ResourceRe
 	}
 	return resources
 }
+
+func (s *ScheduleHandler) updateSchedule() error {
+	if err := s.Client.Update(s.CTX, s.schedule); err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("error updating resource %s/%s: %w", s.schedule.Namespace, s.schedule.Name, err)
+	}
+	return nil
+}
+
+func (s *ScheduleHandler) updateStatus() error {
+	err := s.Client.Status().Update(s.CTX, s.schedule)
+	if err != nil {
+		s.Log.Error(err, "Could not update SyncConfig.", "name", s.schedule)
+		return err
+	}
+	s.Log.Info("Updated SyncConfig status.")
+	return nil
+}
+
+func (s *ScheduleHandler) getOrGenerateSchedule(jobType k8upv1alpha1.JobType, originalSchedule string) string {
+	if s.schedule.Status.GeneratedSchedules == nil {
+		s.schedule.Status.GeneratedSchedules = make(map[k8upv1alpha1.JobType]string)
+	}
+	if schedule, found := s.schedule.Status.GeneratedSchedules[jobType]; found {
+		return schedule
+	}
+	newSchedule := s.randomizeSchedule(s.schedule.Namespace+"/"+s.schedule.Name+"@"+string(jobType), originalSchedule)
+	s.schedule.Status.GeneratedSchedules[jobType] = newSchedule
+	s.requireStatusUpdate = true
+	return newSchedule
+}
+
+// randomizeSchedule randomizes the given originalSchedule with a seed. The originalSchedule has to be one of the supported
+// '@x-random' predefined schedules, otherwise it returns originalSchedule unmodified.
+func (s *ScheduleHandler) randomizeSchedule(seed, originalSchedule string) (schedule string) {
+	hash := sha1.New()
+	hash.Write([]byte(seed))
+	sum := hex.EncodeToString(hash.Sum(nil))
+	sumBase16, _ := new(big.Int).SetString(sum, 16)
+	defer s.Log.V(1).Info("Randomized schedule", "seed", seed, "sum", sum, "from_schedule", originalSchedule, "new_schedule", &schedule)
```

Comment:
IMHO it would make more sense to store `schedule` in a variable and log & return it at the end of the function instead of using `defer` and named returns...

Reply:
Just as a note: using defer for logging, allows to log at the latest possible point in time while not having to take into consideration all error cases in the rest of the function. Here, this feat is not required though.

Reply:
the only reason why that funtion needs the ScheduleHandler was because of a debug message... Now I've somewhat restructured this so that logging happens in the caller (business logic)